microcosm_flask/sync/push.py

```diff
@@ -25,16 +26,39 @@ def push_yaml(inputs, destination):
     safe_dump_all(({href: resource} for href, resource in inputs), destination)
 
 
-def push_json(inputs, base_url, batch_size):
+def push_json(inputs, base_url, batch_size, enable_sessions=False,  max_attempts=2):
     """
     Write inputs to remote URL as JSON.
 
     Future implementations could perform paginated PATCH requests to a collection URI
     if our conventions support it.
 
     """
+    # either use a session or use plain requests
+    session_factory = requests.Session if enable_sessions else lambda: requests
+
+    session = session_factory()
+    for uri, resources in iter_json_batches(inputs, base_url, batch_size):
+        # retry on connection failures
+        for attempt in range(max_attempts):
+            try:
+                if batch_size == 1:
+                    push_resource_json(session, uri, resources[0])
+                else:
+                    push_resource_json_batch(session, uri, resources)
+            except ConnectionError as error:
+                logger.info("Connection error for uri: {}: {}".format(uri, error))
+                # on connection failure, recreate the session
+                session = session_factory()
+                continue
+            else:
+                break
+        if attempt == max_attempts:
```

Comment:
In my tests, the connection resets always recovered after one failure. A down server failed the second time.